### PR TITLE
*: support the vector index

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/55693

Problem Summary:
Merge https://github.com/pingcap/tidb/tree/feature/vector-index to the master branch.
* support vector index at the parser, https://github.com/pingcap/tidb/pull/55698
* executor: Provide TIFLASH_INDEXES table, https://github.com/pingcap/tidb/pull/55901
* support for adding/dropping vector index, https://github.com/pingcap/tidb/pull/55839
* support create table with vector indexes and others, https://github.com/pingcap/tidb/pull/56124 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support the vector index
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: pkg/parser/parser.y [hunk 1]
```diff
@@ -1134,9 +1135,11 @@ import (
 	ConnectionOption                       "single connection options"
 	ConnectionOptionList                   "connection options for CREATE USER statement"
 	ConnectionOptions                      "optional connection options for CREATE USER statement"
-	Constraint                             "table constraint"
+	Constraint                             "table constraint outwith vector index"
```
Comment (tangenta) [inline pkg/parser/parser.y:1138]:
```suggestion
	Constraint                             "table constraint"
```

File: errors.toml [hunk 2]
```diff
@@ -1566,6 +1566,11 @@ error = '''
 Auto analyze is not effective for index '%-.192s', need analyze manually
 '''
 
+["ddl:9014"]
+error = '''
+TiFlash backfill index failed: TiFlash backfill index failed: %s
```
Comment (tangenta) [inline errors.toml:1571]: Why repeat the error message twice?

File: pkg/ddl/executor.go [hunk 3]
```diff
@@ -4586,6 +4589,158 @@ func (e *executor) CreatePrimaryKey(ctx sessionctx.Context, ti ast.Ident, indexN
...
+	if err = checkAddColumnTooManyColumns(len(t.Cols()) + len(hiddenCols)); err != nil {
+		return pmodel.CIStr{}, nil, errors.Trace(err)
+	}
+
+	return indexName, hiddenCols, nil
+}
+
+func checkTableTypeForVectorIndex(tblInfo *model.TableInfo) error {
+	if tblInfo.TableCacheStatusType != model.TableCacheStatusDisable {
+		return errors.Trace(dbterror.ErrOptOnCacheTable.GenWithStackByArgs("Create Vector Index"))
```
Comment (tangenta) [inline pkg/ddl/executor.go:4644]:
```suggestion
		return dbterror.ErrOptOnCacheTable.GenWithStackByArgs("Create Vector Index")
```

Consistent with other errors

File: pkg/parser/parser.y [hunk 4]
```diff
@@ -4221,11 +4234,25 @@ CreateIndexStmt:
...
+     		yylex.AppendError(ErrSyntax)
+     		return 1
+     	}
+     	partSpecs := $10.([]*ast.IndexPartSpecification)
+     	if keyType == ast.IndexKeyTypeVector {
+     		if len(partSpecs) != 1 || partSpecs[0].Expr == nil {
+     			yylex.AppendError(ErrSyntax)
+     			return 1
+     		}
+     	}
```
Comment (tangenta) [inline pkg/parser/parser.y:4249]: Please go to parser directory and execute `make fmt`.

File: pkg/ddl/create_table.go [hunk 5]
```diff
@@ -510,12 +512,47 @@ func checkGeneratedColumn(ctx *metabuild.Context, schemaName pmodel.CIStr, table
...
+		if replicas == 0 {
+			return errors.Trace(dbterror.ErrUnsupportedAddVectorIndex.FastGenByArgs("unsupported TiFlash store count is 0"))
+		}
+
+		// Always try to set to 1 as the default replica count.
+		defaultReplicas := uint64(1)
+		tblInfo.TiFlashReplica = &model.TiFlashReplicaInfo{
+			Count:          defaultReplicas,
+			LocationLabels: make([]string, 0),
+		}
```
Comment (tangenta) [inline pkg/ddl/create_table.go:544]: There are some differences between here and `onSetTableFlashReplica`. Is there any risk?

**Review by AilinKid — APPROVED** [review]

File: pkg/ddl/create_table.go (hunk 5 above)
Comment (zimulala) [reply pkg/ddl/create_table.go:544]: At present, our partition table does not support this feature, so there is no need to check. The check of "checkTiFlashReplicaCount" is also done above. At present, there is no risk.

File: pkg/ddl/create_table.go [hunk 6]
```diff
@@ -538,6 +575,10 @@ func checkTableInfoValidExtra(ec errctx.Context, tbInfo *model.TableInfo) error
 	if err := checkGlobalIndexes(ec, tbInfo); err != nil {
 		return errors.Trace(err)
 	}
+	// A special rule on Serverless is to add TiFlash replica by default if there is a vector index.
```
Comment (tangenta) [inline pkg/ddl/create_table.go:578]:
```suggestion
```

**Review by tangenta — COMMENTED** [review]
Rest LGTM

**Review by tangenta — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**